---
title: "Scarab ddRADseq data analysis"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Demultiplexing and basic quality control 

We've just recieved A LOT (9.39 GB) of data: 100 bp single-end read sequencing data from 239 samples of 11 species of dung beetle (Scarabaeinae), multiplexed on a single Illumina lane, to be precise. Before we dig in to the all the interesting questions we want  to ask with it, we should do a basic check that the sequencing reaction was successful, and that we don't need to start from scratch (lolsob). First, we need to demultiplex the original file, `raw_data/UTN-DUNG_S167_L007_R1_001.fastq.gz1`, using the list of barcodes we appended to each sample during library preparation (`ipyrad/barcodes`). We can look at it and see it matches our expectations for .fastq data, which I won't display here. (Enter all shell commands in the appropriate directory via terminal.)

```{bash, eval=FALSE}
gunzip -c ~/Dropbox/scarab_migration/raw_data/UTN-DUNG_S167_L007_R1_001.fastq.gz | head -n 12
```

We'll initially use the program `ipyrad` [which has extensive documentation you can view here](https://ipyrad.readthedocs.io/) to demultiplex this file into our original libraries. We first generate a params file.

```{bash, eval=FALSE}
ipyrad -n scarab
```

We modify the file `ipyrad/params-scarab.txt` to match our data and fill in the appropriate paths to our data and barcode file. We then run the first step of the ipyrad only, which we indicate with the `-s` option:

```{bash, eval=FALSE}
ipyrad -p params-scarab.txt -s 1
```

When this completes, you'll see the relevant directory fill with fastq files, and a text file labeled `s1_demultiplex_stats.txt` generated. We want to extract data on the total number of reads per library, and group libraries by species. First, we'll read in the (unformatted) stats file and select only the information we want:

```{r}
setwd("~/Dropbox/scarab_migration/ipyrad/")
file <- "s1_demultiplex_stats.txt"
nsamples <- 239
raw <- readLines("s1_demultiplex_stats.txt",warn=FALSE)
a <- grep("sample_name",raw) # where the data begins
a <- a[1] 
reads <- read.table(file,skip=(a),nrow=(nsamples))
colnames(reads) <- c("sample_ID", "read_count")
```

We know have a simple two column data frame to play around with. We will do some tedious substitution of our field-based morphospecies assignment shorthand with the real species IDs:

```{r}
reads$sample_ID <- gsub("EU","Eurysternus_affin",reads$sample_ID)
reads$sample_ID <- gsub("DI1","Dichotomious_satanas",reads$sample_ID)
reads$sample_ID <- gsub("DI5","Dichotomious_satanas",reads$sample_ID)
reads$sample_ID <- gsub("DI7","Dichotomious_podalirius",reads$sample_ID)
reads$sample_ID <- gsub("DE2","Deltochilum_tesselatum",reads$sample_ID)
reads$sample_ID <- gsub("DE1","Deltochilum_speciosissimum",reads$sample_ID)
reads$sample_ID <- gsub("DE9","Deltochilum_speciosissimum",reads$sample_ID)
reads$sample_ID <- gsub("239","Dichotomious_protectus",reads$sample_ID)
reads$sample_ID <- gsub("238","Dichotomious_protectus",reads$sample_ID)
reads$sample_ID <- gsub("219","Deltochilum_amazonicum",reads$sample_ID)
reads$sample_ID <- gsub("220","Coprophanaeus_telamon",reads$sample_ID)
reads$sample_ID <- gsub("217","Phanaeus_meleagris",reads$sample_ID)
reads$sample_ID <- gsub("218","Phanaeus_meleagris",reads$sample_ID)
reads$sample_ID <- gsub("223","Oxysternon_silenus",reads$sample_ID)
reads$sample_ID <- gsub("222","Oxysternon_silenus",reads$sample_ID)
reads$sample_ID <- gsub("221","Oxysternon_conspicullatum",reads$sample_ID)
reads$sample_ID <- gsub('-.*', '', reads$sample_ID) # drop numbers
colnames(reads) <- c("Species","Read_count")
reads$Read_count <- as.numeric(as.character(reads$Read_count))
```

Kind of a mess, but that's biology! Now we'll plot it.

```{r, echo=FALSE}
library(ggplot2)
ggplot(reads, aes(x=Species, y=Read_count)) +
  theme_classic() +
  geom_boxplot()+
  geom_jitter(height = 0, width = 0.1) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, size=9)) +
  ylab("Number of reads")
```

So far so good: pretty even sequencing effort across the five species we're interested in with the most data. 

## Data assembly

Now that we're comfortable we sequenced *something*, we're going to try and assemble the remainder of our data. We're only going to analyze the five species with reasonable sampling: *Deltochilum speciocissimum*, *Deltochilum tesselatum*, *Dichotomius podalirius*, *Dichotomius satanas*, and *Eurysternus affin*. We'll run the rest of the assembly and variant  calling steps using [Jon Puritz' dDocent pipeline](https://www.ddocent.com/), which is optimized for nonmodel organisms and does a good job maximizing the signal to noise ratio in tricky denovo assemblies. First, we'll make separate directories and move our files (named by their initial morphospecies ID codes) to them. We'll then use a bash oneliner paired with files linking our current sample names to a modified sample name format dDocent requires to get ready to run the pipeline. Finally, we'll run dDocent using a set of config files for each species; these are largely the same, except for a more lenient clustering threshhold for  *D. satanas* based on exploratory data analysis.

```{bash, eval=FALSE}

# make directories
mkdir d_speciocissimum
mkdir d_tesselatum
mkdir d_podalirirus
mkdir d_satanas
mkdir e_affin

# move files
mv DE1* d_speciocissimum/
mv DE9* d_speciocissimum/ # two morphospecies lumped into one species
mv DE2* d_tesselatum/
mv DI7* d_podalirius/ 
mv DI1* d_satanas/
mv DI5* d_satanas/
mv EU* e_affin/ 

# rename files
cd d_speciocissimum/; eval "$(sed 's/^/mv /g' d_spec_rename.txt )"; cd ../.
cd d_tesselatum/; eval "$(sed 's/^/mv /g' d_tess_rename.txt )"; cd ../.
cd d_podalirius/; eval "$(sed 's/^/mv /g' d_pod_rename.txt )"; cd ../.
cd d_satanas/; eval "$(sed 's/^/mv /g' d_satanas_rename.txt )"; cd ../.
cd e_affin/; eval "$(sed 's/^/mv /g' e_affin_rename.txt )"; cd ../.

# run independently;
cd d_speciocissimum/; dDocent d_spec_config.file; cd ../.
cd d_tesselatum/; dDocent d_tess_config.file; cd ../.
cd d_podalirius/; dDocent d_pod_config.file; cd ../.
cd d_satanas/; dDocent d_satanas_config.file; cd ../.
cd e_affin/; dDocent e_affin_config.file; cd ../.
```

The output we're interested in are .vcf files; we'll be working with these for the rest of the analysis. First, though, we'll run a set of standard filtering steps on the .vcf files for each species. We're going to drop any individuals with missing data at more than 30% of loci, any loci missing data at more than 25% of individuals, all SNPs with a minimum minor allele frequency of 0.05 and a minimum minor allele count of 3, all SNPs with a quality (Phred) score of <30, and any genotypes with fewer than 5 reads across all individuals. We'll then run dDocent's `dDocent_filter` command, which further trims for allelic balance, depth, and other parameters you can read about in the [program's documentation](https://www.ddocent.com/filtering/). All are very sensible. I've put all my commands together in a simple bash script, `snp_filtering.sh`, but let the buyer beware: it's written to be a record of what I did, and not reproducible, although it would be easy enough to tweak. For now, we'll run it as I did, with a single command:

```{bash, eval=FALSE}
bash scripts/snp_filtering.sh
```

Download these from your cluster, or whereever you're running your analyses. From here on out, we'll be working in `R`.

## Plotting sampling localities

Before we start analyzing our genotypic data in R, it will be useful to visualize where we collected our samples. We'll do this using the `ggmap` package and our coordinate data in `data/scarab_spp_master.csv`. We'll also set up a common color palette to use for the remainder of our study. First, we'll load our data, and use a Google Maps API to grab contour maps of our field sites. 

```{r, message=FALSE}
library(ggmap)
library(wesanderson)
library(gridExtra)

# read in locality data
localities <- read.csv("/Users/ethanlinck/Dropbox/scarab_migration/data/scarab_spp_master.csv")
levels(localities$locality) # see levels 
localities$transect <- ifelse(grepl("CC",localities$locality),'colonso','pipeline') # add levels for transects

# pick the center point for our maps, and an appropriate zoom level
pipeline <- get_map(location=c(lon=-77.85, lat=-0.615), zoom = 13, color = "bw")
colonso <- get_map(location=c(lon=-77.89, lat=-0.937), zoom = 14, color = "bw")

# subset locality data by transect
df.pipeline <- localities[localities$transect=='pipeline',]
df.colonso <- localities[localities$transect=='colonso',]
```


We'll use the "Darjeeling1" palette from the [wesanderson package](https://github.com/karthik/wesanderson), but build a custom palette so we can hold colors constant across the levels of our sampling localities. 


```{r}
pal <- wes_palette("Darjeeling1", 14, type = "continuous")
scale_color_wes <- function(...){
    ggplot2:::manual_scale(
        'color', 
        values = setNames(pal, levels(localities$short_locality)), 
        ...
    )
}
```

Now, we'll plot these using [ggmap](https://github.com/dkahle/ggmap). 

```{r, echo=FALSE}
p1 <- ggmap(pipeline) +
  theme_bw() +
  geom_text(data = df.pipeline,
             aes(x = long, y = lat, color = short_locality, label = short_locality)) +
  scale_color_wes() +
  ggtitle("Pipeline transect") +
  theme(axis.title = element_blank(), legend.position = "none", plot.margin = unit(c(0,0,0,0), "lines"))

p2 <- ggmap(colonso) +
  theme_bw() +
  geom_text(data = df.colonso,
             aes(x = long, y = lat, color = short_locality, label = short_locality)) +
  scale_color_wes() +
  ggtitle("Colonso de Chalupas transect") +
  theme(axis.title = element_blank(), legend.position = "none", plot.margin = unit(c(0,0,0,0), "lines"))

grid.arrange(p1,p2,ncol=2)
```

Not quite publication quality, but good enough for now. 

## Testing for genetic differentiation

Now that we're somewhat spatially oriented, we'll dive in to analyzing our genotypic data. Our driving question is whether dispersal (and gene flow) is reduced across elevational gradients relative to within an elevational band. We are going to try and answer this question in a number of ways. An important first step is running principle component analysis to ask whether there is population genetic structure, and whether it relates to elevation. We'll read in locality data, our .vcf files, and run principle component analysis on each species in turn. This will involve redundant chunks of code. First, *Dichotomius satanas*, our species with the most throrough sampling (*n*=100).

```{r, echo=TRUE,message=FALSE}
# load libraries
library(vcfR)
library(adegenet)
library(ggplot2)
library(wesanderson)
library(tidyverse)
library(reshape2)
library(patchwork)

# set wd and read in locality data
localities <- read.csv("/Users/ethanlinck/Dropbox/scarab_migration/data/scarab_spp_master.csv")

### d. satanas ###

# read in .vcf and sample / pop names; convert to genlight
satanas.vcf <- read.vcfR("/Users/ethanlinck/Dropbox/scarab_migration/raw_data/d_satanas_filtered.FIL.recode.vcf", verbose = FALSE)

# pca 
satanas.dna <- vcfR2DNAbin(satanas.vcf, unphased_as_NA = F, consensus = T, extract.haps = F)
satanas.gen <- DNAbin2genind(satanas.dna)
satanas.pops <-  gsub( "_.*$", "", rownames(satanas.gen@tab))
satanas.gen@pop <- as.factor(satanas.pops)
satanas.scaled <- scaleGen(satanas.gen,NA.method="zero",scale=F)
satanas.pca <- prcomp(satanas.scaled,center=F,scale=F)
satanas.pc <- data.frame(satanas.pca$x[,1:3])
satanas.pc$sample <- rownames(satanas.pc)
satanas.pc$pop <- satanas.pops
satanas.pc$species <- rep("Dichotomius_satanas",nrow(satanas.pc))
```

We've run PCA and merged the first three PC axes with our sampling data. Let's plot PC1 and PC2, and also plot PC1  against elevation. (Note that the colors will jump around during these initial plots, but will match our  sampling locality map once  we've merged the different data frames. So fear not!)

```{r, echo=FALSE}
# merge with sample data
satanas.pc <- merge(satanas.pc, localities, by.x = "sample", by.y = "ddocent_ID")

a <- ggplot(data=satanas.pc,aes(x=PC1,y=PC2,col=satanas.pc$short_locality)) + 
  theme_classic() +
  geom_point()+
  scale_color_manual(values=pal,name="locality")

b <- ggplot(data=satanas.pc,aes(x=elevation,y=PC1,col=satanas.pc$short_locality)) + 
  theme_classic() +
  geom_point()+
  scale_color_manual(values=pal,name="locality")

a + b

```

Looks like panmixia to me, but otherwise reasonable, with no wild outliers. We'll continue as before for the remaining four species. Next up is *Deltochilum speciocissimum*. 

```{r, echo=TRUE,message=FALSE}
# read in .vcf and sample / pop names; convert to genlight
spec.vcf <- read.vcfR("/Users/ethanlinck/Dropbox/scarab_migration/raw_data/d_spec_filtered.FIL.recode.vcf", verbose = FALSE)

# pca 
spec.dna <- vcfR2DNAbin(spec.vcf, unphased_as_NA = F, consensus = T, extract.haps = F)
spec.gen <- DNAbin2genind(spec.dna)
spec.pops <-  gsub( "_.*$", "", rownames(spec.gen@tab))
spec.gen@pop <- as.factor(spec.pops)
spec.scaled <- scaleGen(spec.gen,NA.method="mean",scale=F)
spec.pca <- prcomp(spec.scaled,center=F,scale=F)
spec.pc <- data.frame(spec.pca$x[,1:3])
spec.pc$sample <- rownames(spec.pc)
spec.pc$pop <- spec.pops
spec.pc$species <- rep("Deltochilum_speciocissimum",nrow(spec.pc))

# merge with sample data
spec.pc <- merge(spec.pc, localities, by.x = "sample", by.y = "ddocent_ID")
```

Plots:

```{r, echo=FALSE}

c <- ggplot(data=spec.pc,aes(x=PC1,y=PC2,col=spec.pc$short_locality)) + 
  theme_classic() +
  geom_jitter()+
  scale_color_manual(values=pal,name="locality")

d <- ggplot(data=spec.pc,aes(x=elevation,y=PC1,col=spec.pc$short_locality)) + 
  theme_classic() +
  geom_point()+
  scale_color_manual(values=pal,name="locality")

c + d
```

Looks reasonable. This time, we see strong popualtion genetic structure that largely segregates across the Cosanga river. A curious exception is a sample at the 2150 locality, which might represent a recent migrant. Now, we'll look at *Deltochilum tesselatum*.

```{r, echo=TRUE,message=FALSE}
# read in .vcf and sample / pop names; convert to genlight
tess.vcf <- read.vcfR("/Users/ethanlinck/Dropbox/scarab_migration/raw_data/d_tess_filtered.FIL.recode.vcf", verbose = FALSE)

# pca 
tess.dna <- vcfR2DNAbin(tess.vcf, unphased_as_NA = F, consensus = T, extract.haps = F)
tess.gen <- DNAbin2genind(tess.dna)
tess.pops <-  gsub( "_.*$", "", rownames(tess.gen@tab))
tess.gen@pop <- as.factor(tess.pops)
tess.scaled <- scaleGen(tess.gen,NA.method="mean",scale=F)
tess.pca <- prcomp(tess.scaled,center=F,scale=F)
tess.pc <- data.frame(tess.pca$x[,1:3])
tess.pc$sample <- rownames(tess.pc)
tess.pc$pop <- tess.pops
tess.pc$species <- rep("Deltochilum_tesselatum",nrow(tess.pc))

# merge with sample data
tess.pc <- merge(tess.pc, localities, by.x = "sample", by.y = "ddocent_ID")
```

Plots: 

```{r, echo=FALSE}

e <- ggplot(data=tess.pc,aes(x=PC1,y=PC2,col=tess.pc$short_locality)) + 
  theme_classic() +
  geom_jitter()+
  scale_color_manual(values=pal,name="locality")

f <- ggplot(data=tess.pc,aes(x=elevation,y=PC1,col=tess.pc$short_locality)) + 
  theme_classic() +
  geom_point()+
  scale_color_manual(values=pal,name="locality")

e + f
```

Panmixia again, although there are fewer samples. Now, *Dichotomius podalirius*. 

```{r, echo=TRUE,message=FALSE}

# read in .vcf and sample / pop names; convert to genlight
pod.vcf <- read.vcfR("/Users/ethanlinck/Dropbox/scarab_migration/raw_data/d_pod_filtered.FIL.recode.vcf", verbose = FALSE)

# pca 
pod.dna <- vcfR2DNAbin(pod.vcf, unphased_as_NA = F, consensus = T, extract.haps = F)
pod.gen <- DNAbin2genind(pod.dna)
pod.pops <-  gsub( "_.*$", "", rownames(pod.gen@tab))
pod.gen@pop <- as.factor(pod.pops)
pod.scaled <- scaleGen(pod.gen,NA.method="mean",scale=F)
pod.pca <- prcomp(pod.scaled,center=F,scale=F)
pod.pc <- data.frame(pod.pca$x[,1:3])
pod.pc$sample <- rownames(pod.pc)
pod.pc$pop <- pod.pops
pod.pc$species <- rep("Dichotomius_podalirius",nrow(pod.pc))

# merge with sample data
pod.pc <- merge(pod.pc, localities, by.x = "sample", by.y = "ddocent_ID")
```

Plots: 

```{r, echo=FALSE}
g <- ggplot(data=pod.pc,aes(x=PC1,y=PC2,col=pod.pc$short_locality)) + 
  theme_classic() +
  geom_jitter()+
  scale_color_manual(values=pal,name="locality")

h <- ggplot(data=pod.pc,aes(x=elevation,y=PC1,col=pod.pc$short_locality)) + 
  theme_classic() +
  geom_point()+
  scale_color_manual(values=pal,name="locality")

g + h
```

Again, a big mess. Note that we have fewer sampling localities here, and that they are from the Colonso de Chalupas gradient. We'll now run our final species, *Eurysternus affin*, which is also from Colonso de Chalupas. 

```{r, echo=TRUE,message=FALSE}
# read in .vcf and sample / pop names; convert to genlight
affin.vcf <- read.vcfR("/Users/ethanlinck/Dropbox/scarab_migration/raw_data/e_affin_filtered.FIL.recode.vcf", verbose = FALSE)

# pca 
affin.dna <- vcfR2DNAbin(affin.vcf, unphased_as_NA = F, consensus = T, extract.haps = F)
affin.gen <- DNAbin2genind(affin.dna)
affin.pops <-  gsub( "_.*$", "", rownames(affin.gen@tab))
affin.gen@pop <- as.factor(affin.pops)
affin.scaled <- scaleGen(affin.gen,NA.method="mean",scale=F)
affin.pca <- prcomp(affin.scaled,center=F,scale=F)
affin.pc <- data.frame(affin.pca$x[,1:3])
affin.pc$sample <- rownames(affin.pc)
affin.pc$pop <- affin.pops
affin.pc$species <- rep("Eurysternus_affin",nrow(affin.pc))

# merge with sample data
affin.pc <- merge(affin.pc, localities, by.x = "sample", by.y = "ddocent_ID")
```

Plots: 

```{r, echo=FALSE}

i <- ggplot(data=affin.pc,aes(x=PC1,y=PC2,col=affin.pc$short_locality)) + 
  theme_classic() +
  geom_jitter()+
  scale_color_manual(values=pal,name="locality")

j <- ggplot(data=affin.pc,aes(x=elevation,y=PC1,col=affin.pc$short_locality)) + 
  theme_classic() +
  geom_point()+
  scale_color_manual(values=pal,name="locality")

i + j
```

This is interesting: two extreme outliers. Let's figure out what they are.

```{r}
# drop outliers
drop <- affin.pc[which(affin.pc$PC1>5),] 
drop <- drop$sample
drop
```

Turns out these two samples were marked as potentially belonging to a species other than *E. affin*. Seems like that's almost certainly the case, so let's drop them from the dataset, and merge all our separate data frames.

```{r, echo=TRUE,message=FALSE}
affin.gen <- affin.gen[indNames(affin.gen)!=drop[1] & indNames(affin.gen)!=drop[2]]
affin.pops <-  gsub( "_.*$", "", rownames(affin.gen@tab))
affin.gen@pop <- as.factor(affin.pops)
affin.scaled <- scaleGen(affin.gen,NA.method="mean",scale=F)
affin.pca <- prcomp(affin.scaled,center=F,scale=F)
affin.pc <- data.frame(affin.pca$x[,1:3])
affin.pc$sample <- rownames(affin.pc)
affin.pc$pop <- affin.pops
affin.pc$species <- rep("Eurysternus_affin",nrow(affin.pc))

# merge with sample data
affin.pc <- merge(affin.pc, localities, by.x = "sample", by.y = "ddocent_ID")

# merge all data frames
master.df <- rbind.data.frame(satanas.pc,spec.pc,tess.pc,pod.pc,affin.pc)
```

Now' let's plot all these together with faceting.

```{r,echo=FALSE}
### plot all with faceting

k <- ggplot(data=master.df,aes(x=elevation,y=PC1,col=master.df$short_locality)) + 
  theme_classic() +
  geom_jitter()+
  scale_color_manual(values=pal,name="locality") +
  facet_wrap(~species.x, scales="free")

k

```

This still looks a lot like panmixia, but this will be useful going forward, as we know we won't be violating model assumptions for future analyses (at least except for *D. speciocissimum*, which we'll have to deal with later). 

